File: aon-occam/src/main/java/com/esferalia/aon/occam/api/model/registry/Project.java
package com.esferalia.aon.occam.api.model.registry;

import java.io.Serializable;
import java.util.Date;

import com.esferalia.aon.occam.api.model.Domain;
import com.esferalia.aon.occam.api.model.project.ProjectHolder;
import com.esferalia.aon.occam.api.model.project.ProjectType;

public class Project implements Serializable {

	private static final long serialVersionUID = -5945149996376642583L;
	
	private Integer id;
	private Domain domain;
	private ProjectType type;
	
	private Registry registry;
	private String name;
	private String alias;
	private Date date;
	private boolean tas;
	private boolean commercial;
	private boolean reservation;
	private boolean active;
	
	private ProjectHolder projectHolder;

	private boolean dirty;
	
	public Integer getId() {
		return id;
	}
	
	public Project setId(Integer id) {
		this.id = id;
		return this;
	}
	
	public Domain getDomain() {
		if(this.domain == null) {
			this.domain = new Domain();
		}
		return domain;
	}
	
	public Project setDomain(Domain domain) {
		this.domain = domain;
		return this;
	}
	
	@Deprecated
	public Integer getProjectTypeId() {
		return getType().getId();
	}

	@Deprecated
	public Project setProjectTypeId(Integer projectTypeId) {
		getType().setId(projectTypeId);
		return this;
	}
	
	@Deprecated
	public String getProjectTypeName() {
		return getType().getDescription();
	}
	
	@Deprecated
	public Project setProjectTypeName(String projectTypeName) {
		getType().setDescription(projectTypeName);
		return this;
	}
	
	public String getName() {
		return name;
	}
	
	public Project setName(String name) {
		setDirty(true);
		this.name = name;
		return this;
	}
	
	public String getAlias() {
		return alias;
	}
	
	public Project setAlias(String alias) {
		setDirty(true);
		this.alias = alias;
		return this;
	}
	
	public Date getDate() {
		return date;
	}
	
	public Project setDate(Date date) {
		setDirty(true);
		this.date = date;
		return this;
	}
	
	public boolean isTas() {
		return tas;
	}
	
	public Project setTas(boolean tas) {
		setDirty(true);
		this.tas = tas;
		return this;
	}
	
	public boolean isCommercial() {
		return commercial;
	}
	
	public Project setCommercial(boolean commercial) {
		setDirty(true);
		this.commercial = commercial;
		return this;
	}
	
	public boolean isReservation() {
		return reservation;
	}
	
	public Project setReservation(boolean reservation) {
		setDirty(true);
		this.reservation = reservation;
		return this;
	}
	
	public boolean isActive() {
		return active;
	}
	
	public Project setActive(boolean active) {
		setDirty(true);
		this.active = active;
		return this;
	}
	
	public Registry getRegistry() {
		if(registry == null) {
			registry = new Registry();
		}
		return registry;
	}
	
	public Project setRegistry(Registry registry) {
		setDirty(true);
		this.registry = registry;
		return this;
	}
	
	public ProjectType getType() {
		if(type == null) {
			type = new ProjectType();
		}
		return type;
	}
	
	public Project setType(ProjectType type) {
		setDirty(true);
		this.type = type;
		return this;
	}
	
	public ProjectHolder getProjectHolder() {
		if(this.projectHolder == null) {
			this.projectHolder = new ProjectHolder();
		}
		return projectHolder;
 	}
	
	public Project setProjectHolder(ProjectHolder projectHolder) {
		setDirty(true);
		this.projectHolder = projectHolder;
		return this;
	}
	
	public boolean isDirty() {
		return dirty;
	}
	
	public Project setDirty(boolean dirty) {
		this.dirty = dirty;
		return this;
	}
	
	protected <T extends Project> T copy(Project project, T child) {
		if (registry != null) {
			child.setId(project.getId());
			child.setDomain(project.getDomain());
			child.setName(project.getName());
			child.setAlias(project.getAlias());
			child.setRegistry(project.getRegistry());
			child.setDate(project.getDate());
			child.setType(project.getType());
			child.setActive(project.isActive());
			child.setCommercial(project.isCommercial());
			child.setReservation(project.isReservation());
			child.setTas(project.isTas());
			child.setDirty(project.isDirty());
		}
		return child;
	}
	
	public void setValues(Project project) {
		setId(project.getId());
		setDomain(project.getDomain());
		setName(project.getName());
		setAlias(project.getAlias());
		setRegistry(project.getRegistry());
		setDate(project.getDate());
		setType(project.getType());
		setActive(project.isActive());
		setCommercial(project.isCommercial());
		setReservation(project.isReservation());
		setTas(project.isTas());
		setDirty(project.isDirty());
		setProjectHolder(project.getProjectHolder());
	}
	
	
	public boolean isEmpty() {
		return getId() == null && getDomain().getId() == null
			&& getType().isEmpty() && getRegistry().isEmpty()
			&& getName() == null && getAlias() == null
			&& getDate() == null;
	}
	
}

File: aon-solutions/src/main/webapp/models/project/Project.js
import { Domain } from '../Domain.js';
import { Registry } from '../registry/Registry.js';
import { ProjectActivity } from './ProjectActivity.js';
import { ProjectHolder } from './ProjectHolder.js';
import { ProjectType } from './ProjectType.js';

export class Project {
    id;
    domain;
    type;
    registry;
    name;
    alias;
    date;
    tas;
    commercial;
    reservation;
    active;
    projectHolder;
    projectHolders;
    projectActivities;
    dirty;

    constructor(project) {
        if(project) {
            this.id = project.id;
            this.domain = new Domain(project.domain);
            this.type = new ProjectType(project.type);
            this.registry = new Registry(project.registry);
            this.name = project.name;
            this.alias = project.alias;
            this.date = project.date;
            this.tas = project.tas || false;
            this.commercial = project.commercial || false;
            this.reservation = project.reservation || false;
            this.projectHolder = new ProjectHolder(project.projectHolder);
            this.projectHolders = project.projectHolders || undefined;
            this.projectActivities = project.projectActivities || [];
            this.active = project.active || true;
            this.dirty  = project.dirty || false;
        } else {
            this.domain = new Domain();
            this.type = new ProjectType();
            this.registry = new Registry();
            this.name = '';
            this.alias = '';
            this.date = new Date();
            this.tas = false;
            this.commercial = false;
            this.reservation = false;
            this.active = true;
            this.projectHolder = new ProjectHolder();
            this.dirty = false;
            this.projectActivities = [];
        }
    }

    getId() {
        return this.id;
    }

    setId(id) {
        this.id = id;
        return this;
    }

    getDomain() {
        return this.domain;  
    }

    setDomain(domain) {
        this.domain = new Domain(domain);
        return this;
    }

    getType() {
        return this.type;
    }

    setType(type) { 
        this.setDirty(true);
        this.type = new ProjectType(type);
        return this;
    }

    getRegistry() {
        return this.registry;
    }

    setRegistry(registry) {
        this.setDirty(true);
        this.registry = registry;
        return this;
    }

    getProjectHolder() {
        return this.projectHolder;
    }

    setProjectHolder(projectHolder) {
        this.setDirty(true);
        this.projectHolder = new ProjectHolder(projectHolder);
        return this;
    }

    getName() {
        return this.name;
    }

    setName(name) {
        this.setDirty(true);
        this.name = name;
        return this;
    }

    getAlias() {
        return this.alias;
    }

    setAlias(alias) {
        this.setDirty(true);
        this.alias = alias;
        return this;
    }

    getDate() {
        return this.date;
    }

    setDate(date) {
        this.setDirty(true);
        this.date = date;
        return this;
    }

    getProjectHolders() {
        return this.projectHolders || [];
    }

    setProjectHolders(projectHolders) {
        this.projectHolders = projectHolders.map(holder => new ProjectHolder(holder).setDirty(true));
        return this;
    }

    addProjectActivity(projectActivity) {
        let opt = this.findProjectActivityByActivityType(projectActivity);
        if(opt){
            opt.removed = false;
        } else {
            this.projectActivities.push(new ProjectActivity(projectActivity).setDirty(true));
        }
    }
    
    removeProjectActivity(projectActivity) {
        let opt = this.findProjectActivityByActivityType(projectActivity);
        if(opt){
            opt.removed = true;
        }
    }

    getProjectActivities() {
        return this.projectActivities;
    }

    setProjectActivities(projectActivities) {
        this.setDirty(true);
        this.projectActivities = projectActivities.map(d => new ProjectActivity(d).setDirty(true));
        return this;
    }

    findProjectActivityByActivityType({activityType}){
        return this.getProjectActivities().find(d=> d.activityType && d.activityType.id === activityType.id);
    }
    
    isTas() {
        return this.tas;
    }

    setTas(tas){
        this.setDirty(true);
        this.tas = tas;
        return this;
    }
    
    isCommercial() {
        return this.commercial;
    }

    setCommercial(commercial){
        this.setDirty(true);
        this.commercial = commercial;
        return this;
    }
    
    isReservation() {
        return this.reservation;
    }

    setReservation(reservation){
        this.setDirty(true);
        this.reservation = reservation;
        return this;
    }

    isActive() {
        return this.active;
    }

    setActive(active){
        this.setDirty(true);
        this.active = active;
        return this;
    }

    isDirty() {
        return this.dirty || this.getProjectHolder().isDirty();
    }

    setDirty(dirty) {
        this.dirty = dirty;
        return this;
    }

    /**
     * 
     * @returns Project
     */
    clone() {
        return new Project(this);
    }
}
